common/lib/xmodule/xmodule/modulestore/django.py

```diff
@@ -308,6 +308,193 @@ def fetch_disabled_xblock_types():
 _MIXED_MODULESTORE = None
 
 
+class CachingModuleStoreWrapper:
+    """
+    Wrap a modulestore and add request-level caching of get_course()
+
+    This class proxies most of its requests to its underlying Modulestore
+    object. It was necessary to do this as a separate wrapping layer because
+    our _MIXED_MODULESTORE can be a CCXModulestoreWrapper or a MixedModuleStore
+    depending on our settings. Putting this caching into MixedModuleStore itself
+    doesn't work because of the way CCXModulestoreWrapper pulls back results
+    from MixedModuleStore and changes their usage keys.
+
+    So yes, this means that content access is layered behind proxies like:
+
+      CachingModuleStoreWrapper ->
+        (CCXModulestoreWrapper -> if CCX enabled)
+          MixedModuleStore ->
+            DraftVersioningModuleStore (Split Mongo)
+            DraftModuleStore (Old Mongo)
+
+    The main thing we have to worry about in this class is to be able to detect
+    any call that might potentially invalidate the cache course, and to do that
+    invalidation before sending the call along.
```

Comment:
This worries me - feels like we're creating a bug for ourselves in a few years, when someone adds a new method to `DraftVersioningModuleStore` but forgets to (or never knew to) add that here as well.

There's no easy way to wrap something like a `save()` call being done, so we know to invalidate the cache? Or how about at least a comment in the code or docstring for the underlying classes, so someone would hopefully know to come edit this class?

Reply:
Yeah, I've been trying to balance the foot-guns of this being self-contained vs. the accuracy of having the modulestores themselves do the cache invalidation (but now having to be aware of the caching). I honestly cringe at the idea of adding something more to the modulestore constructors, but thinking over your comment, I think that would be the safest thing overall.

So maybe the wrapper caches, while the modulestores manually call invalidation code right next to their database writes. Something like:

1. Make a course cache object that's backed by the RequestCache and smart enough to know how to invalidate multiple keys by course.
2. Pass that course cache object through to the constructors of all the Modulestores. The only thing they should be doing with it is invalidation. Then invalidation can happen only at the inner-most layer where the actual database writes are occurring.
3. Pass the same course cache object to `CachingModuleStoreWrapper`. Still make the high level `CachingModuleStoreWrapper` do the actual caching and return-from-cache behavior, because doing so from the lower levels would break with CCX's boxing/unboxing of usage keys. This also keeps the change safer, since we can toggle this entirely by a Django settings and roll it back quickly in case of a regression detected on prod.

This is only kind of related to my current TNL work (hence the weekend hacking). I noticed while doing some development last week that modulestore access overhead had grown on certain endpoints and threatened to negate some of the work we're trying to do with speeding up the courseware, so I wanted to dig in and see if there was a not-crazy fix.

Reply:
This is a slightly different vibe, but maybe you could avoid making the Modulestores know anything about this specific cache by having them just grow a new method that the wrapper overrides?

Something like `commit_to_store` or `do_side_effects_of_a_write` or something. (This only works if the modulestore is a parent class of the wrapper, but I think that's how this is all set up yeah?)

Reply:
Yeah, along those lines, rather than wrapping the modulestore, can you put this cache into `get_course_by_id()` itself? And what about using split's `update_course_index()` as the single point of cache invalidation? Every change which you've identified here in this wrapper ultimately involves updating the course index to a new version, so that it points to the new structure with the change. Or does the bulk operations somehow make that not useful in some cases?

Reply:
@mikix:

> This is a slightly different vibe, but maybe you could avoid making the Modulestores know anything about this specific cache by having them just grow a new method that it calls on write and that the wrapper overrides?
>
> Something like commit_to_store or do_side_effects_of_a_write or something. (This only works if the modulestore is a parent class of the wrapper, but I think that's how this is all set up yeah?)

The wrapper object that CCX uses is not part of the Modulestore class hierarchy at all–it's just a plain subclass of object that knows how to proxy through most calls anything would make on it.

I'd prefer to keep the new thing clear of the Modulestore class hierarchy as well, since it's already pretty complicated. I think calling a method on a new cache obj to invalidate a course is about on par with calling a new method to do side-effects, in terms of how much a Modulestore is forced to understand?

> Yeah, along those lines, rather than wrapping the modulestore, can you put this cache into get_course_by_id() itself?

That was the first version of this PR, but I found that some of the worst offenders have a reference to the modulestore and call `get_course()` on it directly. It also didn't seem like it added that much complexity to put it a dumb wrapper vs. in that module–I basically copied the CCX wrapper verbatim for the proxy bits, and that's been in the system for a long time. I do think that the long line of manually proxied methods for side-effects is bad, but...

> And what about using split's update_course_index() as the single point of cache invalidation?

...but I think that this would work very well to cover the invalidations. So in this scenario, the caching wrapper only implements the generic pass-throughs (which is already being done with CCX's wrapper object), and the exact methods for things it's actually caching, which is just `get_course` in this case. For split, we'd put the cache invalidation right in `update_course_index`, and do something like the equivalent for Old Mongo as well.